lib/endpoints/vnc.js

```diff
@@ -0,0 +1,304 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var jsprim = require('jsprim');
+var restify = require('restify');
+var schemas = require('joyent-schemas').cloudapi;
+var util = require('util');
+var vasync = require('vasync');
+var watershed = require('watershed');
+var mooremachine = require('mooremachine');
+var net = require('net');
+
+var shed = new watershed.Watershed();
+
+function mount(server, before, pre) {
+    assert.object(server, 'server');
+    assert.ok(before, 'before');
+    assert.optionalArrayOfFunc(pre, 'pre');
+
+    pre = pre || [];
+
+    server.get({
+        path: '/:account/machines/:machine/vnc',
+        name: 'ConnectMachineVNC',
+        version: [ '8.4.0' ]
+    }, before, connectVNC);
+
+    return server;
+}
+
+function connectVNC(req, res, next) {
+    var vm = req.machine;
+    if (vm.brand !== 'kvm') {
+        res.send(400, new restify.RestError({
+            statusCode: 400,
+            restCode: 'MachineHasNoVNC',
+            message: 'Specified machine does not have a VNC console'
+        }));
+        next();
+        return;
+    }
+    if (vm.state !== 'running') {
+        res.send(400, new restify.RestError({
+            statusCode: 400,
+            restCode: 'MachineStopped',
+            message: 'Specified machine is stopped and cannot be connected to'
+        }));
+        next();
+        return;
+    }
+    if (!res.claimUpgrade) {
+        res.send(400, new restify.RestError({
+            statusCode: 400,
+            restCode: 'UpgradeRequired',
+            message: 'VNC connect endpoint is a websocket and must be Upgraded'
+        }));
+        next();
+        return;
+    }
+
+    /*
+     * Since cloudapi still runs with restify request domains enabled, we need
+     * to exit that domain here if we want any errors in the VNC FSM to be
+     * reported sensibly (since the request will end from restify's
+     * perspective once we send the 101).
+     *
+     * This can be removed once domains and the uncaughtException handler are
+     * turned off for cloudapi.
+     */
+    var reqdom = process.domain;
+    reqdom.exit();
```

Comment:
I would try to future proof this - by adding a check against whether a domain actually exists, maybe:
```
if (process.domain && process.domain.domain) {
}
```

Note that `process.domain` is not documented at all - this is what it shows me on the node.js CLI:
```
> process.domain
Domain {
  domain: null,
  _events: { error: [Function] },
  _eventsCount: 1,
  _maxListeners: undefined,
  members: [] }
> process.domain.domain
null
```

Reply:
Since I'm just trying to shepherd this in and don't know about the details here w/ CloudAPI, I'm happy to defer to you here